### ISSUE TITLE
Changing NumberInput does not set form dirty

### ISSUE THREAD

**Issue by lvernaillen** [marmelab/react-admin#7919]
**What you were expecting:**
Changing a `NumberInput` sets the form dirty and enables the save button.
Similar to other input types as `TextInput`, `ReferencInput`, `DateInput`, ...
When saving, the number shown in the `NumberInput` is saved.

**What happened instead:**
Changing a `NumberInput` does not set the form dirty and does not enable the save button.
Whether you change a `NumberInput` by typing in it, by using keyboard up/down arrow or by clicking on the up/down arrows in the input, it doesn't make a difference.
You need to leave the input first, make it loose focus. Only then the save button gets enabled and the value is bound to that input.

This is especially problematic in a scenario with `submitOnEnter` enabled, which is the default.
If the last input you change is that `NumberInput` and you press Enter, then the **old** value of the `NumberInput` is saved. The changed `NumberInput` value is not saved!

**Steps to reproduce:**
- Use your default code sandbox
  https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple

Old values are saved:

- Go to the posts resource
- Click on a post to go to the edit view
- Choose the Miscellaneous tab
- Modify the Category (save button gets enabled)
- Click inside the "Average note" input
- Modify that "Average note" value using keyboard or up/down arrows
- Don't leave that field but simply press Enter to save the form
- Notice the Category is changed but the "Average note" value is not!

Save button not enabled:

- Go to the posts resource
- Click on a post to go to the edit view
- Choose the Miscellaneous tab
- Modify the "Average note"
- Notice the save button not being enabled untill you leave the input

**Other information:**
We currently work around this by disabling the submitOnEnter functionality  (as described in the [upgrade docs](https://marmelab.com/react-admin/Upgrade.html#submitonenter-prop-has-been-removed)) to prevent old values being saved.
But even without `submitOnEnter`, it is not user friendly needing an extra click to be able to save a changed `NumberInput`.
One click outside the `NumberInput` to enable the save button and one click to do the actual save.
In Firefox 102.0 (64-bit) on Win10 21H2 (OS build 19044.1766), removing focus from the `NumberInput` still does not enable the save button when input is changed using the arrows in the input itself.

**Environment**

* React-admin version: 4.2.0
* Last version that did not exhibit the issue (if applicable):
* React version: 17.0.2
* Browser: Chrome Version 103.0.5060.66 (Official Build) (64-bit)

**Comment by slax57** [comment]
Hi @lvernaillen 
Thank you for submitting this.

The NumberInput validating its value only after loosing focus is actually on purpose, see [this PR](https://github.com/marmelab/react-admin/pull/7673) for more explanations.

However, I agree that the NumberInput should also update its value and validate it when we press the enter key.
This is most certainly something we can fix.
I'm labeling this as a bug.

**Comment by lvernaillen** [comment]
Hi @slax57,

Updating the value and validating when pressing the enter key helps in scenario where you first modified another field which enabled the save button. However when the only value you change is that Numberinput, then the save button isn't enabled. In the latter case you need to press Enter twice. Once to blur the NumberInput which enables the save button and a second time to actually save the form.

There is also the Firefox case where, if you change the NumberInput by clicking directly on the up/down arrows in the input field, the value is never updated, even when trying to blur by clicking outside the input. You really need to click on an area inside the NumberInput that is not the up/down arrow area and then outside the NumberInput to blur it (and have the value updated).
Should I create a separate issue for this Firefox bug
![numberinput-firefox-not-updated](https://user-images.githubusercontent.com/4180374/178502656-d0255042-57b8-42fa-aa6c-22cc54f27d1f.gif)
?

**Comment by lvernaillen** [comment]
Screen capture of issue where removing focus from the NumberInput still does not enable the save button when input is changed using the arrows in the input itself in Firefox (Chrome is ok).

![Animation5](https://user-images.githubusercontent.com/4180374/178509350-a65cc974-8def-4097-88dc-afc8481b509a.gif)

Firefox 102.0 (64-bit) on Win10 21H2 (OS build 19044.1766)

**Comment by bladx** [comment]
Sorry for the duplicated ticket.

As I said in my ticket, the issue is not present on https://marmelab.com/react-admin-demo/#/products/create/details, so Maybe that's a clue on how to solve this.


> This is a Firefox bug: the browser doesn't trigger a focus or a blur event when using the arrow keys on a number input (see https://jsbin.com/fekuceyohi/edit?html,console,output for a simple demo). Again, if this is a problem for you, use use a `<TextInput type="number">` instead of a `<NumberInput>`.

I found out the same problem than Lieven, but was able to reproduce it on both Firefox and Google Chrome.
I believe this ticket is way more detailed than my freshly created one (https://github.com/marmelab/react-admin/issues/8028), but maybe it contains new information.
